### PR TITLE
MPSC ring buffer

### PR DESCRIPTION
**Pull request by nsavoire**
# What does this PR do?
Add MPSC ring buffer

Design is based on BPF ring buffer (https://www.kernel.org/doc/html/latest/bpf/ringbuf.html).
    * Writer first reserves some space at ring buffer head by first taking a lock, advancing
      writer position, marking sample as busy and releasing lock. Lock is required because
      advancing position / marking sample cannot be done atomically.
    * Writer commits a sample by atomically setting its state to ready or discarded
    * Reader is almost the same as for Perf ring buffer, it just additionally check sample state:
      if sample is busy it continues waiting, if sample is discarded, it silently drops it

### CODE REVIEW

File: include/ringbuffer_utils.hpp
Comment (ivoanjo) [inline include/ringbuffer_utils.hpp:169]: Minor: I think this can be slightly simplified:

```suggestion
    update_tail();
```

File: include/ringbuffer_utils.hpp
Comment (ivoanjo) [inline include/ringbuffer_utils.hpp:258]: Minor: 

```suggestion
  inline size_t available_size() const { return _head - _tail; }

  ConstBuffer read_sample() {
    size_t n = available_size();
```

File: src/perf_ringbuffer.cc
Comment (ivoanjo) [inline src/perf_ringbuffer.cc:34]: I'm kinda curious about this one -- why store pointers to the `MPSCRingBufferMetaDataPage` struct, rather than putting that struct directly inside `struct RingBuffer`? (potentially as a union, to support the multiple ring buffer types)

File: test/ringbuffer-ut.cc
Comment (ivoanjo) [inline test/ringbuffer-ut.cc:263]: Minor: I had to admit I had to look up [emplace](https://www.macmillandictionary.com/dictionary/british/emplace) in the dictionary 😅

**Review by ivoanjo — COMMENTED** [review]
Hey! @r1viollet suggested I may learn a bit from this PR, and indeed I did!

I also left a few n00b questions, hopefully they aren't too bad :)

(I'm also more scared of C++ than I was before!)

File: include/mpscringbuffer.hpp
Comment (r1viollet) [reply include/mpscringbuffer.hpp:46]: My understanding is that the order of arrival of the threads and 500 microseconds might allow the threads not to go back to spinning. Though perhaps in some apps all threads arrive at the same time.

File: include/mpscringbuffer.hpp
Comment (r1viollet) [inline include/mpscringbuffer.hpp:29]: should we consider a second safety mechanism to get out ?

File: include/ringbuffer_utils.hpp
Comment (sanchda) [inline include/ringbuffer_utils.hpp:165]: I think I understand.  In the common setting, the underlying RingBuffer is going to be shared--either in a `mmap(...MAP_SHARED);` or in a separate thread (multiprocess is the worst-case, so I'll focus on that).  On the other hand, the MPSCRingBufferWriter/Reader classes are _views_ into the Ringbuffer, with all private variables intended to be local to the current thread.

File: include/ringbuffer_utils.hpp
Comment (sanchda) [inline include/ringbuffer_utils.hpp:169]: Got it.  Instantiate the view by taking a snapshot of the current tail.  There's no reason to have a view (or in initialization) of the head of the ringbuffer, since that needs to be sequenced carefully at time of reservation.

File: include/ringbuffer_utils.hpp
Comment (sanchda) [inline include/ringbuffer_utils.hpp:205]: Right!  Even though we're locked, we need to push down the write to the _next_ reader, in case the write doesn't go through before we unlock.

File: include/mpscringbuffer.hpp
Comment (sanchda) [inline include/mpscringbuffer.hpp:41]: Since this spinlock is also supposed to support multi-process coordination, I think we might need to make it more robust.  Here are some random ideas.

1.  As Ivo notes in the other comment, when `n_waiters >> n_schedulable_cores`, then it becomes more likely for the thread holding the lock to slow down in a way that compounds contention.  The way that some of the pthreads types handle this is by having the slow-path do an atomic increment indicating the number of "interested" (haha) threads.  Unfortunately, the pthreads code I checked only uses this field for arithmetic, but I guess we could use a combination of the lock-defined `num_interested` and thread-local `interested_idx` to sequence the waits?  This isn't a complete solution, just an idea.
2.  As Erwan indicated, since we're at the mercy of other processes, which may be killed in various situations (user culling heavy processes, OOM, or even in the case where the lock is held across container boundaries and an unrelated service gracefully closes), I wonder if we need to structure a check during the slow path.  This is actually really hard to do because the lock-holder may be in a separate PID namespace.  I'd say that after a certain number of passes (plus the addition of a sequence number or something, so we don't get a lot of threads thrashing on the lock), one of the threads needs to make a call to either take ownership of the lock or declare the lock invalid so the application can figure out what to do.

Lots of words with nothing specific, sorry :(

File: include/mpscringbuffer.hpp
Comment (sanchda) [reply include/mpscringbuffer.hpp:46]: This type of fast-path/slow-path spinlock is surprisingly durable.  I'd structure a test which sits on a tight loop over the lock in a pool of threads, doubling the number of threads every iteration from 1x to 1024x the number of CPUs and observe the scaling properties of the lock itself after a token amount of on-CPU work.  I'd expect
* Median wait time per thread to remain linear up to a big multiple of the number of cores.
* After breaking linear scaling, we get quadratic rather than exponential

Why do I think this?  Jiffies are big and CPUs are fast.  All bets are off the table if the work loop has a context switch though, in that case I'm with Ivo.

But again, I think the point is to look at the scaling properties rather than a given a priori overhead value for a fixed number of threads.

I do think that ceiling should be evaluated though.  We're adding a spinlock to a library which is trying to infect its children, and some processes are more, uh, prodigious breeders than others.  :)

File: include/ringbuffer_utils.hpp
Comment (sanchda) [inline include/ringbuffer_utils.hpp:179]: I really like `reserve()` semantics here.

One of the challenges though is that a consumer doesn't know if a given element is ready, only if the whole ringbuffer is ready.  I think the current strategy for that problem is good, since the alternative is to have a bitmask for potential reader positions indicating whether the given read position is ready (since we have variable-size elements).  I don't think that's a favorable alternative unless we start hitting issues with sparse updates (lagging thread in the middle of the ringbuffer) or high write contention with fast drains (in which case the buffer is never full but always busy), but I very much doubt that will happen given how big our samples are.

File: include/ringbuffer_utils.hpp
Comment (sanchda) [inline include/ringbuffer_utils.hpp:185]: I also don't know whether it's the best, but I do think it's quite lovely.  My $0.02 USD (using USD right now because it's weaker than the Euro and my opinion shouldn't be worth much) is that any decisions about different lock architecture or modifying the reservation sequence to use CAS+atomics depends entirely on the usage pattern, and for that we need to deploy what we have and observe it in its natural setting.

File: include/ringbuffer_utils.hpp
Comment (sanchda) [inline include/ringbuffer_utils.hpp:216]: Beautiful.  Again, I love reservation semantics.

File: include/ringbuffer_utils.hpp
Comment (sanchda) [inline include/ringbuffer_utils.hpp:243]: Hey, is head/tail read-after-write sequencing an issue here?  I see that we set the MO for the head, and I see that the reserve/commit sequence by all means *should* update the head before the tail.  My gut tells me that this is fine.  Any reason to believe otherwise?

File: include/ringbuffer_utils.hpp
Comment (sanchda) [inline include/ringbuffer_utils.hpp:269]: Makes sense to me.  Leave it to the calling application to determine how to handle the scenario where the ringbuffer is in the middle of a reserve-commit sequence.

File: test/ringbuffer-ut.cc
Comment (sanchda) [reply test/ringbuffer-ut.cc:263]: lol, also https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2008/n2642.pdf

But fundamentally, the C++ standards committee has a precise technical goal for using the term "emplace" to qualify a particular type of insertion.

**Review by sanchda — DISMISSED** [review]
Some pretty good stuff here!  I like the spinlock, I think it's going to scale pretty well.  I like the reserve-commit semantics.  I think we should all do a little bit of scaling analysis, but I don't think that work needs to block merging into main.

My only reservation is that this is a multiprocess MPCS structure, but because of its application domain we don't have a mechanism to detect lock-holder expiry/deadlock.  I think we can structure a solution, but I would be very happy making that improvement a subsequent PR rather than overloading this one with too many expectations.

File: include/mpscringbuffer.hpp
Comment (sanchda) [reply include/mpscringbuffer.hpp:46]: IMHO, this is pretty idiomatic.
1. Use of pause/yield on x86/arm to tell the processor that this is a spin-wait loop
2. Use of OS sched yield to sleep thread after fixed number of iterations in the fast path, back to lock

One of the problems with the classic spinlock is, as Ivo notes, suppose that the spinlock is the only thing on CPU, with a bunch of on- and off-cpu threads in contention.  Draining the spinlock will depend on the active thread being able to progress, and whenever it has to yield to the scheduler it might have to round-robin behind all of the waiting threads as they clash on the lock.

File: test/ringbuffer-ut.cc
Comment (ivoanjo) [reply test/ringbuffer-ut.cc:263]: Someone wrote 8 pages on this matter....

/me backs away slowly and silently

File: include/mpscringbuffer.hpp
Comment (nsavoire) [reply include/mpscringbuffer.hpp:13]: C++ standards are often not fully implemented (even when compilers claim support), especially on the standard library side !
clang has not support yet for hardware_destructive_interference_size yet (https://libcxx.llvm.org/Status/Cxx17.html) and only recent gcc versions (>=12.1) have support for it (https://gcc.gnu.org/onlinedocs/libstdc++/manual/status.html#status.iso.2017)

File: src/perf_ringbuffer.cc
Comment (nsavoire) [reply src/perf_ringbuffer.cc:34]: I hesitated some time between both solutions. I think at the start I thought I could factorize code between ring buffer implementations by having a common pointer that expose the same semantic but points to different parts depending on the type.
But I agree that since I was not able to factorize code, clearly separating the data for different implementations might make the code mode readable.

File: test/ringbuffer-ut.cc
Comment (nsavoire) [reply test/ringbuffer-ut.cc:263]: C++ is too complicated for its own good !
The idea behind `emplace` is to be able to construct an object in-place in a container, potentially saving the construction / copy of a temporary object.

**Comment by ivoanjo** [comment]
Thanks for the explanations @nsavoire! Learned a bunch of modern C++ things with this PR :)

File: include/mpscringbuffer.hpp
Comment (nsavoire) [reply include/mpscringbuffer.hpp:29]: I see @r1viollet's point:  if a thread/process dies while having the lock, then other processes will wait endlessly.
I will add a timeout.

File: include/mpscringbuffer.hpp
Comment (nsavoire) [reply include/mpscringbuffer.hpp:32]: Not my idea ;)
https://probablydance.com/2019/12/30/measuring-mutexes-spinlocks-and-how-bad-the-linux-scheduler-really-is/

File: include/ringbuffer_utils.hpp
Comment (nsavoire) [reply include/ringbuffer_utils.hpp:165]: The object model is not clear because there is not object that represents the ring buffer and is responsible for its creation/destruction.
`RingBuffer` holds the data necessary to access the ring buffer but does not act on it. The only requirement is a zero-initialized chunk of memory. And it is indeed shared explicitly (several threads accessing the same object) or implicitly (pointers in different `RingBuffer` objects pointing to the same underlying memory).
Writer/Reader are not really views (IMHO `RingBuffer` is the view) but local short-lived objects that modify the state of the ring buffer in a consistent manner.

File: include/ringbuffer_utils.hpp
Comment (nsavoire) [reply include/ringbuffer_utils.hpp:205]: Indeed the lock is only used by writers, reader/writer synchronization relies on atomic loads/stores.

File: include/mpscringbuffer.hpp
Comment (ivoanjo) [reply include/mpscringbuffer.hpp:46]: Ahhhh you're right about `spincount`, I completely missed that part and read it as the loop starting as-if from scratch 😅

File: include/mpscringbuffer.hpp
Comment (nsavoire) [reply include/mpscringbuffer.hpp:41]: Concerning point 1, I will add a benchmark to be able to base our decisions on measures.
Point 2 is very important, I definitely need to add safeguards. Disabling allocation profiling when we detect that a lock is never released seems the safest choice to me.

File: include/ringbuffer_utils.hpp
Comment (nsavoire) [reply include/ringbuffer_utils.hpp:269]: For the calling application, we behave as if the ringbuffer was empty, we don't convey the information that a sample is being finalized.
We could try reading the next samples but this would add a lot more complexity. WDYT ?

File: include/ringbuffer_utils.hpp
Comment (nsavoire) [reply include/ringbuffer_utils.hpp:243]: I am not sure to understand your concern.
`_tail` is only written by reader and read by writer, reads of `_tail` in reader can be relaxed since writes are done by the same thread.

File: include/ringbuffer_utils.hpp
Comment (nsavoire) [reply include/ringbuffer_utils.hpp:179]: I would be interested to discuss your idea !
Initially I started an implementation without a spinlock, the idea was to use a ready bit, at the start the ready bit would be unset, writer would reserve memory by advancing head through a CAS operation (as it is done here), but reader would not read the sample since its ready is not set yet. Upon commit writer would do an exchange on the ready bit to set it, and reader would do an exchange on it with 0 to check if sample is ready and at the same time reset it to not ready when ready. Sadly this does not work with variable-size samples :(
The other idea was to have a "pre-head" that ahead of head, reserving would advance the pre-head through CAS. Here there is not need for ready/busy state, but commit is more complex because only the writer that has reserved the sample just after head is allowed to advance head, which leads to a wait loop.

**Comment by nsavoire** [comment]
Thanks @sanchda, @ivoanjo and @r1viollet for all your comments !

File: include/ringbuffer_utils.hpp
Comment (nsavoire) [inline include/ringbuffer_utils.hpp:236]: Ooops, `relaxed` is not good here, since we need that all previous writes to be visible to the reader once we change `hdr->size`. Need to be changed to `release`

**Review by r1viollet — DISMISSED** [review]
LGTM
An impressive PR ! Thanks for splitting this up

**Review by r1viollet — APPROVED** [review]
LGTM

**Review by r1viollet — DISMISSED** [review]
LGTM

**Review by r1viollet — APPROVED** [review]
LTGM -- Thanks for adding extra tests on forking

**Pull request merged**
